File: Camera_ctrl.py
# camera_ctrl - provides basic control of QSI RS 6.1s CCD camera
#
# Matthew Grossman from Princeton HCIL - Jun. 5, 2018
# Based on a Matlab version developed by He Sun
#
#Brief Usage:
#    Turn on the camera and enable the handle:
#        h = Camera_ctrl(0, 'enable')
#    Initilialization:
#        Camera_ctrl(h, 'init', temperature) # temperature range: (-50, 50)
#    Stop fan and cooler:
#        Camera_ctrl(h, 'finalize')
#    Disable the handle:
#        Camera_ctrl(h, 'finalize')
#    Open or close shutter:
#        h = Camera_ctrl(h, 'shutter', openflag) 
#        # openflag: True for open, False for close
#    Set readout speed: Camera_ctrl(h, 'readoutspeed', readoutflag)
#        # readout flag: 1 for fast readout, 0 for high image quality
#    Exposure properties:
#        Camera_ctrl(h, 'exposureproperties', start_pos, size_pixels,
#                    bin_pixels)
#            # the last three inputs are all 2 tuples with default values of 
#            # (0,0), (2758, 2208), and (1,1)
#    Set shutter priority: Camera_ctrl(h, 'shutterpriority', shutterflag)
#        # shutterflag: 0 for mechanical, 1 for electrical
#    Take pictures: img = Camera_ctrl(h, 'exposure', exptime)
#        # exptime is the exposure time you can change
#    Show camera realtime picture: Camera_ctrl(h, 'realtime')
#        # can be used during calibration

from matplotlib import pyplot as plt
import numpy
import win32com.client

class Camera:
    """ A class that mimics the handle class from for cameras matlab and
    initializes to default values of serialnum: 0, shutter: open, and default
    starting position, pixel size, and bin pixels to (0,0)
       """
    def __init__(self):
        self.handle = None
        self.serialnum = '0'
        self.shutter = True
        self.startPos = (0, 0)
        self.imgSize = (0, 0)
        self.binPix = (0, 0)
        self.ccdtemp = None

def Camera_ctrl(camera, cmd, *args):
    
    n_argin = len(args)
    cmd = cmd.lower()
    
    if cmd == 'connect':
        try:
            # get a handle of the camera
            if camera.handle == None:
                camera.handle = win32com.client.Dispatch('QSICamera.CCDCamera')
            # connect the camera
            if camera.handle.Connected == False:
                camera.camera.Connected = True
                print('Connecting ' + camera.handle.__class__.__name__)
            else:
                print(camera.handle.__class__.__name__ + 'is already connected.')
            # get camera dfault peramters
            camera.serialnum = camera.handle.SerialNumber;
            camera.defaultsizepixels = (camera.handle.Numx, camera.handle.NumY)
            camera.defaultstartpos = (0, 0)
            camera.defaultbinpixels = (1, 1)
            camera = Camera_ctrl(camera, 'shutter', True)
        except Exception as ex:
            pass
#########
            ######## ask Christian about finding the names for these exceptions
        return camera
    
    elif cmd == 'disconnect':
        if camera.handle == None:
            raise Exception('Camera not connected.')
        # close shutter
        camera = Camera_ctrl(camera, 'shutter', False)
        Camera_ctrl(camera, 'finalize')
        Camera_ctrl(camera, 'disable')
        print('Disconnecting ' + camera.handle.__class__.__name__)
        camera.handle = None
        return camera
    
    elif cmd == 'init':
        if camera.handle == None:
            raise Exception('Camera not connected.')
        if n_argin != 1:
            raise ValueError('Wrong number of input arguments')
        # sets the current camera as the main camera
        if camera.handle.IsMainCamera == False:
            camera.handle.IsMainCamera = True
        # turns on the camera fan
        if camera.handle.FanMode != 'FanFull':
            camera.handle.FanMode = 'FanFull'
        # enable the CCD cooler
        if camera.handle.CoolerOn != True:
            camera.handle.CoolerOn = True
        # set camera cooling temperature and update ccdtemp attribute
        ccdtempc = args[0]
        if camera.handle.CanSetCCDTemperature == True:
            camera.handle.SetCCDTemperature = ccdtempc
            camera.ccdtemp = camera.handle.SetCCDTemperature
        # set camera gain to low gain
        if camera.handle.CameraGain != 'CameraGainLow':
            camera.handle.CameraGain = 'CameraGainLow'
            # set camera shutter priority to electrical
            # 0 for mechanical, 1 for electical
            camera.handle.ShutterPriority = 1
            return camera
            
    elif cmd == 'shutter':
        if camera.handle == None:
            raise Exception('Camera not connected.')
        if n_argin != 1:
            raise ValueError('Wrong number of input arguments')
        openflag = args[0]
        if openflag == True:
            # Set the camera to manual shutter mode.
            camera.handle.ManualShutterMode = True
            # Open the shutter as specified
            camera.handle.ManualShutterOpen = True
            camera.shutter = True
        else:
            # Set the camera to manual shutter mode
            camera.handle.ManualShutterMode = True
            # Close the shutter as specified
            camera.handle.ManualShutterOpen = False
            # Set the camera to auto shutter mode
            camera.handle.ManualShutterMode = False
            camera.shutter = False;
        return camera
    
    elif cmd == 'exposureproperties':
        if camera.handle == None:
            raise Exception('Camera not connected.')
        if n_argin != 3:
            raise ValueError('Wrong number of input arguments')
        startPos = args[0]
        imgSize = args[1]
        binPix = args[2]
        if len(startPos) != 2:
            raise ValueError('Wrong dimension of start position')
        if len(imgSize) != 2:
            raise ValueError('Wrong dimension of picture size')  
        if len(binPix) != 2:
            raise ValueError('Wrong dimension of binned pixels')
        # sends the exposure properties to the camera
        camera.handle.StartX = startPos[0]
        camera.handle.StartY = startPos[1]
        camera.handle.NumX = imgSize[0]
        camera.handle.NumY = imgSize[1]
        camera.handle.BinX = binPix[0]
        camera.handle.BinY = binPix[1]
        # update the camera attributes
        camera.startPos = (camera.handle.StartX, camera.handle.StartY)
        camera.imgSize = (camera.handle.NumX, camera.handle.Numy)
        camera.binPix = (camera.handle.BinX, camera.handle.BinY)
        return camera
    
    elif cmd == 'avgimg':
        if camera.handle == None:
            raise Exception('Camera not connected.')
        if n_argin != 2:
            raise ValueError('Wrong number of input arguments')
        camera.handle.ReadoutSpeed = camera.fastReadout
        expTime = args[0]
        numIm = args[1]
        img = numpy.zeros((camera.imgSize[1], camera.imgSize[0]), numpy.int32)
        for i in range(numIm):
            img = img + Camera_ctrl(camera, 'exposure', expTime)
        return img / numIm
    
    elif cmd == 'exposure':
        if camera.handle == None:
            raise Exception('Camera not connected.')
        if n_argin != 1:
            raise ValueError('Wrong number of input arguments')
        exptime = args[0]
        # Starts an exposure on the camera
        camera.shutter.StartExposure(exptime, camera.shutter)
        # Wait for the exposure to complete
        done = camera.handle.ImageReady
        while done != True:
            done = camera.handle.ImageReady
#HOW WILL THE SAFEARRAY COME OUT?
        # gets the image from the camera as some form of array
        return camera.handle.ImageArray
    
    elif cmd == 'realtime':
        # creates the figure
        plt.figure(100)
        plt.title('Real Time Picture')
        plt.tight_layout()
        # updates the figure with the current image until the figure is closed
        while plt.fignum_exists(100):
            img = Camera_ctrl(camera, 'exposure', 0.0003)
            plt.imshow(img)
            plt.pause(.1)
        
    elif cmd == 'readoutspeed':
        if camera.handle == None:
            raise Exception('Camera not connected.')
        if n_argin != 1:
            raise ValueError('Wrong number of input arguments')
        readoutflag = args[0]
        # sends the readout speed to the camera
        camera.handle.ReadoutSpeed = readoutflag      
                
    elif cmd == 'shutterpriority':
        if camera.handle == None:
            raise Exception('Camera not connected.')
        if n_argin != 1:
            raise ValueError('Wrong number of input arguments')
        shutterflag = args[0]
        # sends the shutter pririty to the camera
        camera.handle.ShutterPriority = shutterflag

    elif cmd == 'finalize':
        if camera.handle == None:
            raise Exception('Camera not connected.')
        # turn off the camera fan
        if camera.handle.FanMode != 'FanOff':
            camera.handle.FanMode = 'FanOff'
        # disable the CCD cooler
        if camera.handle.CoolerOn != False:
            camera.handle.CoolerOn = False
        # closes the shutter
        Camera_ctrl(camera, 'shutter', False)
        camera.shutter = False
        
    elif cmd == 'disable':
        if camera.handle == None:
            raise Exception('Camera not connected.')
        # disconnects camera
        if camera.handle.Connected == True:
            camera.handle.Connected = False
    else:
        print('unkown command: ' + cmd)

File: Camera_ctrl.m
% camera_ctrl - provides basic control of QSI RS 6.1s CCD camera
%
% He Sun from Princeton HCIL - Sep. 4, 2015
% Updates from Christian Delacroix - 2018
%
% Brief useage:
%   Connect the camera:
%           camera.name = 'QSICamera.CCDCamera';
%           camera = Camera_ctrl(camera, 'connect');
%   Disconnect the camera:
%           camera = Camera_ctrl(camera, 'disconnect');
%   Initialization:
%           Camera_ctrl(camera, 'init', ccdtemp); % temperature: range -50 to
%           50
%   Stop fan and cooler:
%           Camera_ctrl(camera, 'finalize');
%   Disable the camera:
%           Camera_ctrl(camera, 'disable');
%   Open or close shutter;
%           camera = Camera_ctrl(camera, 'shutter', openflag); % openflag: 1 for open
%           0 for close
%   Set readout speed:
%           Camera_ctrl(camera, 'readoutspeed', readoutflag); % readoutflag: 1
%           for fast readout, 0 for high image quality
%   Exposure properties:
%           camera = Camera_ctrl(camera, 'exposureproperties', startPos, imgSize,...
%                       binPix); % the last three input arguments are
%                       all 2 dimension vectors, the default values are
%                       [0,0], [2758,2208], [1,1]
%   Set shutter priority:
%           Camera_ctrl(camera, 'shutterpriority', shutterflag); % shutterflag:
%           0 for mechanical, 1 for electrical
%   Take pictures:
%           img = Camera_ctrl(camera, 'exposure', expTime); % exposure time in seconds
%   Show camera realtime picture:
%           Camera_ctrl(camera, 'realtime'); % can be used during calibration

function output = Camera_ctrl(camera, cmd, varargin)

    n_argin = size(varargin,2);
    switch(lower(cmd))
        
        case 'connect'
            try 
                % get a handle of the camera
                if not(isfield(camera, 'handle'))
                    camera.handle = actxserver(char(camera.name));
                end
                % connect the camera
                if not(camera.handle.Connected)
                    camera.handle.Connected = 1;
                    disp(['Connecting ' class(camera.handle)])
                else
                    disp([class(camera.handle) ' is already connected.'])
                end
                % get camera default parameters
                camera.serialnum = camera.handle.SerialNumber;
                camera.defaultsizepixels = [camera.handle.NumX, camera.handle.NumY];
                camera.defaultstartpos = [0,0];
                camera.defaultbinpixels = [1,1];
                camera = Camera_ctrl(camera, 'shutter', 1); % open shutter
            catch ME
                switch ME.identifier
                    case 'MATLAB:COM:InvalidProgid'
                        fprintf ( 2, [camera.name +': wrong camera name!\n\n'] )
                        disp([camera.name +': wrong camera name.'])
                    case 'MATLAB:COM:E2147748395'
                        fprintf ( 2, 'Lost handle of camera, unplug USB manually!\n\n' )
                        camera = rmfield(camera, 'handle');
                    otherwise
                        rethrow(ME)
                end
            end
            output = camera;

        case 'disconnect'
            assert(isfield(camera, 'handle'), 'Camera not connected.')
            camera = Camera_ctrl(camera, 'shutter', 0); % close shutter
            Camera_ctrl(camera, 'finalize');
            Camera_ctrl(camera, 'disable');
            disp(['Disconnecting ' class(camera.handle)])
            camera = rmfield(camera, 'handle');
            output = camera;
            
        case 'init'
            assert(isfield(camera, 'handle'), 'Camera not connected.')
            assert(n_argin==1, 'Wrong number of input arguments');
            % set current camera as the main camera
            if not(camera.handle.IsMainCamera)
            	camera.handle.IsMaincamera = 1;
            end
            % turn on the camera fan
            if not(strcmp(camera.handle.FanMode, 'FanFull')) % if FanOff
                camera.handle.FanMode = 'FanFull';
            end
            % enable the CCD cooler
            if not(camera.handle.CoolerOn)
                camera.handle.CoolerOn = 1;
            end
            % set camera cooling temperature, and update ccdtemp attribute
            ccdtemp = cell2mat(varargin(1));
            if camera.handle.CanSetCCDTemperature
                camera.handle.SetCCDTemperature = ccdtemp;
            end
            camera.ccdtemp = camera.handle.SetCCDTemperature;
            % set camera gain to low gain
            if not(strcmp(camera.handle.CameraGain, 'CameraGainLow')) % if CameraGainHigh
                camera.handle.CameraGain = 'CameraGainLow';
            end
            % set camera shutter priority to electrical
            camera.handle.shutterpriority = 1; % 0 for mechanical, 1 for electrical
            output = camera;

        case 'shutter'
            assert(isfield(camera, 'handle'), 'Camera not connected.')
            assert(n_argin==1);
            openflag = cell2mat(varargin(1));
            if  openflag== 1;
                % Set the camera to manual shutter mode
                set(camera.handle, 'ManualShutterMode', 1);
                % Open the shutter as specified
                set(camera.handle, 'ManualShutterOpen', 1);
                camera.shutter = 1;
            else
                % Set the camera to manual shutter mode
                set(camera.handle, 'ManualShutterMode', 1);
                % Close the shutter as specified
                set(camera.handle, 'ManualShutterOpen', 0);
                % Set the camera to auto shutter mode
                set(camera.handle, 'ManualShutterMode', 0);
                camera.shutter = 0;
            end
            output = camera;
            
        case 'exposureproperties'
            assert(isfield(camera, 'handle'), 'Camera not connected.')
            assert(n_argin==3,'Wrong number of input arguments');
            startPos = cell2mat(varargin(1));
            imgSize = cell2mat(varargin(2));
            binPix = cell2mat(varargin(3));
            assert(length(startPos)==2, 'Wrong dimension of start position');
            assert(length(imgSize)==2, 'Wrong dimension of image size');
            assert(length(binPix)==2, 'Wrong dimension of binned pixels');
            %assert((startPos(1)+imgSize(1))*binPix(1)<=h.defaultsizepixels(1), 'x pixels exceeds chip range!');
            %assert((startPos(2)+imgSize(2))*binPix(2)<=h.defaultsizepixels(2), 'x pixels exceeds chip range!');
            camera.handle.StartX = startPos(1);
            camera.handle.StartY = startPos(2);
            camera.handle.NumX = imgSize(1);
            camera.handle.NumY = imgSize(2);
            camera.handle.BinX = binPix(1);
            camera.handle.BinY = binPix(2);
            % update camera attributes
            camera.startPos = [camera.handle.StartX, camera.handle.StartY];
            camera.imgSize = [camera.handle.NumX, camera.handle.NumY];
            camera.binPix = [camera.handle.BinX, camera.handle.BinY];
            output = camera;
            
        case 'avgimg'
            assert(isfield(camera, 'handle'), 'Camera not connected.')
            assert(n_argin==2, 'Wrong number of input arguments');
            camera.handle.readoutspeed = camera.fastReadout;
            expTime = cell2mat(varargin(1));
            numIm = cell2mat(varargin(2));
            img = int32(zeros(camera.imgSize(2), camera.imgSize(1)));
            for itr = 1:numIm
                img = img + Camera_ctrl(camera, 'exposure', expTime);
            end
            output = img / numIm;
            
        case 'exposure'
            assert(isfield(camera, 'handle'), 'Camera not connected.')
            assert(n_argin==1, 'Wrong number of input arguments');
            assert(camera.shutter==1||camera.shutter==0, 'The shutter status is incorrect');
            expTime = cell2mat(varargin(1));
            invoke(camera.handle, 'StartExposure', expTime, camera.shutter);
            % Wait for the exposure to complete
            done = get(camera.handle, 'ImageReady');
            while not(done)
                done = get(camera.handle, 'ImageReady');
            end
            output = get(camera.handle, 'ImageArray')';
            
        case 'realtime'
            f = figure(100);
            set(f, 'Name', 'Real Time Picture');
            %set(camera.handle, 'ShutterPriority', 1);
            while(ishandle(f))
                img = Camera_ctrl(camera, 'exposure', 0.0003);
                figure(100),imagesc(img), axis xy tight, colorbar;
                drawnow
            end
            output = 1;
            
        case 'readoutspeed'
            assert(isfield(camera, 'handle'), 'Camera not connected.')
            assert(n_argin==1,'Wrong number of input arguments');
            readoutflag = cell2mat(varargin(1));
            set(camera.handle, 'readoutspeed', readoutflag);
            output = 1;
        case 'shutterpriority'
            assert(isfield(camera, 'handle'), 'Camera not connected.')
            assert(n_argin==1,'Wrong number of input arguments');
            shutterflag = cell2mat(varargin(1));
            set(camera.handle, 'shutterpriority', shutterflag);
            output = 1;
        case 'finalize'
            assert(isfield(camera, 'handle'), 'Camera not connected.')
            % turn off the camera fan
            temp=get(camera.handle, 'FanMode');
            if strcmpi(temp,'FanOff') ~= 1
                set(camera.handle, 'FanMode', 'FanOff');
            end
            % disable the CCD cooler
            temp=get(camera.handle, 'CoolerOn');
            if temp ~= 0
                set(camera.handle, 'CoolerOn', 0);
            end
            Camera_ctrl(camera, 'shutter', 0);
            output = 1;
        case 'disable'
            assert(isfield(camera, 'handle'), 'Camera not connected.')
            % disconnect camera
            temp=get(camera.handle, 'Connected');
            if temp == 1
                set(camera.handle, 'Connected', 0);
            end
            output = 0;
            
        otherwise
            disp([ 'unknown command ' cmd ]);
            output = -1;
    end
end